logging/proto_conversions.go

```diff
@@ -0,0 +1,69 @@
+package logging
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// FieldKeyAndValueFromProto examines a *structpb.Struct and returns its key
+// string and native golang value.
+func FieldKeyAndValueFromProto(field *structpb.Struct) (string, any, error) {
+	var fieldValue any
+
+	fieldJSON, err := json.Marshal(field)
+	if err != nil {
+		return "", nil, err
+	}
+
+	var zf zap.Field
+	if err := json.Unmarshal(fieldJSON, &zf); err != nil {
+		return "", nil, err
+	}
+
+	if zf.Type == zapcore.ErrorType {
+		fmt.Printf("zf: %v\n", zf)
+	}
+
+	// This code is modeled after zapcore.Field.AddTo:
+	// https://github.com/uber-go/zap/blob/fcf8ee58669e358bbd6460bef5c2ee7a53c0803a/zapcore/field.go#L114
+	//nolint:exhaustive
+	switch zf.Type {
+	case zapcore.BoolType:
+		fieldValue = zf.Integer == 1
+	case zapcore.DurationType:
+		fieldValue = time.Duration(zf.Integer)
+	case zapcore.Float64Type:
+		// Converting floats to proto sometimes introduces small amounts of loss
+		// (e.g. an extra .0000000000073). Use a hacky combination of math, fmt,
+		// and strconv to use only up to 5 decimal places.
```

Comment:
Talked to @zaporter-work offline; 5 decimal places seems somewhat unacceptable. I could increase the number of decimal places, but I'm going to start by taking a whack at _why_ `protoutils.StructToStructpb` loses precision 🕵🏻 .

Reply:
😮‍💨 it's not our code [it's google's `protobuf`](https://pkg.go.dev/google.golang.org/protobuf@v1.32.0/types/known/structpb#NewValue) package that has lossy encoding for `int64` and `uint64`.

> When converting an int64 or uint64 to a NumberValue, numeric precision loss is possible since they are stored as a float64.